File: ruby/luhn/luhn.rb
# Module Luhn comment
module Luhn
  VERSION = 2
  def self.valid?(number)
    return false if invalid_luhn_format?(number)

    (luhn_number(number) % 10).zero?
  end

  private_class_method def self.invalid_luhn_format?(number)
    number.delete!(' ')
    number.size < 2 || number =~ /[A-Za-z]/ || number =~ /\D/
  end

  private_class_method def self.array_luhn(number)
    number.scan(/\d/).reverse.map.with_index do |num, index|
      if index.odd?
        double_num = num.to_i * 2
        double_num > 9 ? double_num -= 9 : double_num
      else
        num.to_i
      end
    end
  end

  private_class_method def self.luhn_number(number)
    array_luhn(number).sum
  end
end

File: go/luhn/luhn.go
package luhn

import (
	"strconv"
	"strings"
)

func Version() string {
	return "1"
}

func Valid(num string) bool {

	// strip spaces from num
	num = strings.Replace(num, " ", "", -1)

	if len(num) <= 1 {
		return false
	}

	// double every digit starting from the right
	// if the product is greater than 9, sum the digits
	// sum all the digits
	// if the sum is divisible by 10, the number is valid
	var sum int
	for i := len(num) - 1; i >= 0; i-- {
		d, error := strconv.Atoi(string(num[i]))

		// if contains a non number
		if error != nil {
			return false
		}

		if (len(num)%2 == 0 && i%2 == 0) || (len(num)%2 == 1 && i%2 == 1) {
			d *= 2
			if d > 9 {
				d -= 9
			}
		}
		sum += d
	}

	return sum%10 == 0
}
